reactivesocket-cpp/src/tcp/TcpDuplexConnection.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include <folly/Memory.h>
+#include <reactivesocket-cpp/src/mixins/MemoryMixin.h>
+#include "TcpDuplexConnection.h"
+
+namespace reactivesocket {
+    using namespace ::folly;
+
+    Subscriber<Payload> &TcpDuplexConnection::getOutput() {
+        if (!outputSubscriber_) {
+            outputSubscriber_ = folly::make_unique<TcpOutputSubscriber>(*this);
+        }
+        return *outputSubscriber_;
+    };
+
+    void TcpDuplexConnection::setInput(Subscriber<Payload> &inputSubscriber) {
+      inputSubscriber_ = &inputSubscriber;
+
+      auto* subscription = new MemoryMixin<TcpSubscriptionBase>();
+
+      inputSubscriber_->onSubscribe(*subscription);
+
+      socket_->setReadCB(this);
+    };
+
+    void TcpDuplexConnection::send(Payload element) {
+      socket_->writeChain(this, std::move(element));
+    }
+
+    void TcpDuplexConnection::writeSuccess() noexcept {
+    }
+
+    void TcpDuplexConnection::writeErr(size_t bytesWritten, const AsyncSocketException& ex) noexcept {
+      std::cout << "TODO writeErr" << bytesWritten << ex.what() << "\n";
+    }
+
+    void TcpDuplexConnection::getReadBuffer(void **bufReturn, size_t *lenReturn) noexcept {
+      std::tie(*bufReturn, *lenReturn) = readBuffer_.preallocate(4096, 4096);
```

Comment:
it is not clear for me whether you are allocating a buffer and still managing the lifetime via IOBufQueue or if you are giving this up. Is readBufferAvailable method giving you the same memory back?